proposal/js-deprecations.d.ts

```diff
@@ -0,0 +1,302 @@
+ *
+ * #### Errors and Warnings for Invalid Deprecations
+ *
+ * For all of the deprecation options, we chose to have the API throw an error
+ * if a string ID is passed for a deprecation that does not exist. This will
+ * ensure that typos are caught early and is consistent with the behavior on the
+ * command line.
+ *
+ * For `fatalDeprecations`, we chose to make passing in a future deprecation
+ * that's not also passed to `futureDeprecations` an error, since, in a
+ * sense, future deprecations aren't really deprecations yet unless they're
+ * enabled via `futureDeprecations`. This also prevents someone who just passes
+ * all deprecations to `fatalDeprecations` from accidentally opting into a
+ * deprecation that hasn't been released yet. This is not yet consistent with
+ * the behavior on the command line / Dart API, but we plan to update them to
+ * match this specification.
+ *
+ * For obsolete deprecations passed to `fatalDeprecations`, we chose to emit a
+ * warning so users can know to update their configuration, but we allow
+ * compilation to proceed without an error, since an obsolete deprecation means
+ * the underlying behavior has now been removed, which should match the behavior
+ * of making that deprecation fatal while it was still active.
+ *
+ * For `silenceDeprecations`, we chose to make passing anything other than
+ * an active deprecation an error to prevent users from unintentionally
+ * over-silencing. Future deprecations are already effectively silenced, so
+ * users shouldn't be able to pass it here until it's actually active. If the
+ * deprecation a user was silencing is now obsolete, that means the breaking
+ * change has now been made, so silencing has no effect, and users who were
+ * silencing it may not be ready for the actual breaking change. There is no
+ * way to distinguish between different user-authored deprecations, so they
+ * should not be able to be silenced as a group.
+ *
+ * For `futureDeprecations`, we only emit a warning for non-future deprecations,
+ * since future deprecations all eventually become active. The warning will let
+ * users know to clean up their configuration, but there should be no behavior
+ * change.
+ */
+
+/* ## API */
+
+import {SourceSpan} from '../spec/js-api';
+
+declare module '../spec/js-api' {
+  interface Options<sync extends 'sync' | 'async'> {
+    /**
+     * A set of deprecations to treat as fatal.
+     *
+     * If a version is provided, then all deprecations released in that
+     * version or earlier will be treated as fatal. If a string is provided, it
+     * will resolve to the deprecation with that ID, or error if none exists.
+     *
+     * This will throw an error if a future deprecation is included here, unless
+     * that future deprecation is also passed to `futureDeprecations`. This
+     * will emit a warning if an obsolete deprecation is included here. */
+    fatalDeprecations?: (Deprecation | string | Version)[];
```

Comment:
You can use `keyof _Deprecation` rather than `string` here to provide better static checking in TypeScript and better autocompletion.